bindings/rascal/utils/fps.py

```diff
@@ -87,3 +93,194 @@ def fps(feature_matrix, n_select, starting_index=None,
+        if n_select is None:
+            n_select = self.Nselect
+
+        if self.act_on in ['sample per species']:
+            sps = list(n_select.keys())
+            # get various info from the structures about the center atom species and indexing
+            (strides_by_sp, global_counter, map_by_manager,
+             indices_by_sp) = get_index_mappings_sample_per_species(managers, sps)
+            selected_ids_by_sp = {key: np.sort(val[:n_select[key]])
+                                  for key, val in self.selected_sample_ids_by_sp.items()}
+            self.selected_ids = convert_selected_global_index2rascal_sample_per_species(
+                managers, selected_ids_by_sp, strides_by_sp, map_by_manager, sps)
+            # return self.selected_ids
+            # build the pseudo points
+            pseudo_points = SparsePoints(self._representation)
+            pseudo_points.extend(managers, self.selected_ids)
+            return pseudo_points
+
+        elif self.act_on in ['sample']:
+            selected_ids_global = np.sort(self.selected_sample_ids[:n_select])
+            strides, _, map_by_manager = get_index_mappings_sample(managers)
+            self.selected_ids = convert_selected_global_index2rascal_sample(managers,
+                                                                            selected_ids_global, strides, map_by_manager)
+            return self.selected_ids
+            # SparsePoints is not compatible with a non center atom species
+            # dependant sparse points
+            # # build the pseudo points
+            # pseudo_points = SparsePoints(self._representation)
+            # pseudo_points.extend(managers, self.selected_ids)
+            # return pseudo_points
+
+        elif self.act_on in ['feature']:
+            feat_idx2coeff_idx = self._representation.get_feature_index_mapping(
+                managers)
+            self.selected_ids = {key: []
+                                 for key in feat_idx2coeff_idx[0].keys()}
+            selected_ids_sorting = np.argsort(
+                self.selected_feature_ids_global[:n_select])
+            selected_feature_ids = self.selected_feature_ids_global[selected_ids_sorting]
+            for idx in selected_feature_ids:
+                coef_idx = feat_idx2coeff_idx[idx]
+                for key in self.selected_ids.keys():
+                    self.selected_ids[key].append(int(coef_idx[key]))
+            # keep the global indices and ordering for ease of use
+            self.selected_ids['selected_features_global_ids'] = selected_feature_ids.tolist(
+            )
+            self.selected_ids['selected_features_global_ids_fps_ordering'] = selected_ids_sorting.tolist(
+            )
+            self.selected_ids = dict(coefficient_subselection=self.selected_ids)
+            return self.selected_ids
+
+    def fit_transform(self, managers):
+        return self.fit(managers).transform(managers)
+
+    def plot(self):
```

Comment:
...are we requiring `matplotlib` as a dependency now? At the moment it's a "stealth dependency" in that the import is only invoked when this function is called, which is one of many good reasons to avoid imports in functions like this. PEP8 is quite clear that all imports should be listed at the top of the module.

As for whether we _should_ require any plotting package as a dependency, I'd rather just return the data to plot and let the user plot it themselves using whatever package they like (even if that is almost always matplotlib).

...but if you do want to have matplotlib-specific functionality, you could do it without the dependency by naming this something like `plot_matplotlib` and taking a matplotlib `Axes` object as an argument instead of using `pyplot`. I'd still advocate for moving plotting and diagnostic tools to a separate utils module, which can be imported or ignored as desired.

Reply:
this is something that a user will have to do most of the times to understand where to 'cut' hence making it into a function. Might be simpler to have matplotlib as a dependency ? Putting this utility into another file feels too much scattering since this function would only work for FPSFilter.

Reply:
Really? I thought this was part of the convergence-testing workflow, where you'd test the error of the fit w.r.t. several convergence parameters of the kernel (including the number of sparse features and points).

It's probably not a huge problem to have Matplotlib as a dependency anyway, since most users will already have it installed (and it's simple and easy to install if not). ASE, for instance, has it as an "optional" dependency, but it's not entirely clear which modules actually depend on it. The best way to handle this is probably to wrap the import (top of file) in a try/catch; the catch part just prints an error that matplotlib isn't installed and that plotting functionality won't be available.

Side note: I'd name this function `plot_fps_error` to make it clear what exactly it's plotting. And, as always, documentation is needed.